tests/templates/template_test.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2021 VMware
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templates
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/vmware-tanzu/cartographer/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/cartographer/pkg/templates"
+	"github.com/vmware-tanzu/cartographer/tests/template-tester"
+)
+
+func TestTemplateExample(t *testing.T) {
+	params, err := template_tester.BuildBlueprintStringParams(template_tester.StringParams{
+		{
+			Name:         "gitops_url",
+			DefaultValue: "https://github.com/vmware-tanzu/cartographer/",
+		},
+		{
+			Name:         "gitops_branch",
+			DefaultValue: "main",
+		},
+	})
+	if err != nil {
+		t.Fatalf("failed to build param: %v", err)
+	}
+
+	workload := createWorkload()
+
+	deliverable := createDeliverable()
+
+	templateOfDeliverable, err := createTemplate(deliverable)
+	if err != nil {
+		t.Fatalf("failed to create template: %v", err)
+	}
+
+	expectedDeliverable := createExpectedDeliverable(deliverable)
+
+	expectedUnstructured := createExpectedUnstructured()
+
+	testSuite := template_tester.TemplateTestSuite{
+		"template, workload and expected defined in files": {
+			Inputs: template_tester.TemplateTestInputs{
+				TemplateFile:    "template.yaml",
+				WorkloadFile:    "workload.yaml",
+				BlueprintParams: params,
+			},
+			Expectations: template_tester.TemplateTestExpectations{
+				ExpectedFile: "expected.yaml",
+			},
+		},
+
+		"template defined as an object": {
+			Inputs: template_tester.TemplateTestInputs{
+				Template:        templateOfDeliverable,
+				BlueprintParams: params,
+				WorkloadFile:    "workload.yaml",
+			},
+			Expectations: template_tester.TemplateTestExpectations{
+				ExpectedFile: "expected.yaml",
+			},
+			IgnoreMetadataFields: []string{"creationTimestamp"},
```

Comment:
A personal feeling, is that I like how [reconciler-runtime lays out the test config](https://github.com/vmware-labs/reconciler-runtime/blob/main/testing/config.go#L33-L77) as a flat list of fields.

For a couple of reasons:
1. It's slightly less cluttered
2. it means that reactors (functions to run during tests) and metadata (ancillary config) etc can be added to the suite as it evolves, without wondering where they belong.

Reply:
I also like the term "Givens" as established state (over "Inputs")

Reply:
I started with everything flat and didn't like it. I DO like the term `Given`. I've changed the struct to use that.